052-record-reconciled-versions.md

```diff
@@ -0,0 +1,185 @@
+
+Example field in status:
+```
+status:
+  versions:
+    lastSuccessfulReconciliationBy: '0.37.0'
+```
+If a reconcile does not reach completion, this field is not added or updated, however in the case where it already exists it is left at the previous value.
+i.e.
+`status.versions.lastSuccessfulReconciliationBy=0.35.0` would remain if a post-operator-upgrade reconciliation on `0.36.0` failed, as the information that a prior reconciliation passed would be useful to an end user.
+
+### status.versions.kafka
+
+The `status.versions.kafka` field is updated in the CR at the end of the KafkaReconciler in the `podsReady` method and signals to a user that the KafkaReconciler has finished reconciling the Kafka brokers to Kafka version `Y`, regardless of success or failure later in the reconcile.
+```
+status:
+  versions:
+    kafka: '3.5.1'
+```
+
+This field is required as it signals that a reconcile to Kafka version `Y` has finished, and this extra information is important and needs to be available, as a user just checking `status.versions.lastSuccessfulReconciliationBy` is not sufficient for upgrades where multiple rolls of the brokers are required, such as a Kafka version update alongside an upgrade.
+
+If the operator errors during the Kafka reconciliation this field will not be updated. 
+If Kafka was successfully updated, the field will also be updated, meaning errors later in the reconcile, such as failing to deploy the `entity-operator` or `kafka-exporter` will still show `status.versions.kafka` correctly up to date, but `lastSuccessfulReconciliationBy` will not be updated, and the CR will no longer be in a `Ready` state.
+
+
+## `Kafka` Custom Resource & Reconciler changes
+
+- Update `KafkaStatus` with new `versions` field and sub fields as agreed on this proposal.
+
+- Update `initialStatus` [method](https://github.com/strimzi/strimzi-kafka-operator/blob/0c4e96de69e20cb80e86fd8ccb3ab8baca431822/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java#L292) to seed new status with the `status.versions` field if it is already set on the CR. This is so that any status update leaves these fields unchanged, unless they are explicitly modified as part of reconcile.
+
+- The update to `status.versions.kafka` would happen after the `podsReady` [method](https://github.com/strimzi/strimzi-kafka-operator/blob/0c4e96de69e20cb80e86fd8ccb3ab8baca431822/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java#L817) updating the Kafka version once all pods are ready. The Kafka version can be determined from the `StrimziPodSet` which has the information available in the `strimzi.io/kafka-version` annotation.
```

Comment:
This seems related to my comment above. Why not store it and update it later when the reconciliation ends (successfully or not)?

Also, why do you need to know the version from the PodSet? The operator has to know the version it uses there, so there should be no need to query the pod set annotation.